File: scripts/ebus-guided/measure-recorded-sector.mjs
/**
 * Measure the lit region of the recorded EBUS clips (EBUS-PRE-REVIEW-02, lane A).
 *
 * The workbench shows `Depth<n>.mp4` letterboxed inside a square box, so the sector occupies a
 * small part of what is rendered. Before cropping anything for display we measured where the
 * recorded content actually is, rather than guessing a border. This script is the record of that
 * measurement: it samples frames across every authored segment window of every depth file and
 * prints the union bounding box of pixels above a luminance floor.
 *
 * The luminance floor and the reporting live in the companion `measure-recorded-sector.py`.
 * It needs ffmpeg and python3 (numpy + Pillow) and is not part of the build or the test run.
 *
 *   node scripts/ebus-guided/measure-recorded-sector.mjs [outputDir]
 */
import { execFileSync } from 'node:child_process'
import { mkdirSync, rmSync } from 'node:fs'
import { join } from 'node:path'
import { tmpdir } from 'node:os'

const DEPTHS = [2, 3, 4, 5, 6, 8]
// Every authored segment start plus a mid-window sample: gain 0–16 s, contrast 16–32 s, flow 32–38 s.
const TIMES = [1, 3, 5, 7, 9, 11, 13, 15, 17, 21, 25, 29, 33, 35, 37]

const mediaDir = 'EBUS-course/apps/web/public/media/knobology/Depth_segments'
const out = process.argv[2] ?? join(tmpdir(), 'ebus-recorded-sector')
mkdirSync(out, { recursive: true })

for (const depth of DEPTHS) {
  for (const time of TIMES) {
    execFileSync('ffmpeg', [
      '-loglevel',
      'error',
      '-ss',
      String(time),
      '-i',
      join(mediaDir, `Depth${depth}.mp4`),
      '-frames:v',
      '1',
      '-y',
      join(out, `d${depth}_t${time}.png`),
    ])
  }
}

execFileSync('python3', [new URL('measure-recorded-sector.py', import.meta.url).pathname, out], {
  stdio: 'inherit',
})
rmSync(out, { recursive: true, force: true })

File: scripts/ebus-guided/measure-recorded-sector.py
"""Union bounding box of lit pixels across sampled frames of the recorded EBUS clips.

Companion to measure-recorded-sector.mjs (EBUS-PRE-REVIEW-02, lane A). Reads the PNG frames that
script extracts and prints, per depth file, where the recorded content actually sits inside the
1920x1080 frame, plus every sampled frame that also carries the device banner.
"""

import glob
import sys

import numpy as np
from PIL import Image

LUMA_FLOOR = 12
BANNER_ABOVE = 60
BANNER_BELOW = 960

per: dict[str, list] = {}
for path in sorted(glob.glob(sys.argv[1] + "/*.png")):
    mask = np.asarray(Image.open(path).convert("L")) > LUMA_FLOOR
    cols, rows = np.where(mask.any(0))[0], np.where(mask.any(1))[0]
    if not len(cols):
        continue
    depth = path.split("/")[-1].split("_")[0][1:]
    per.setdefault(depth, []).append(
        (int(cols[0]), int(rows[0]), int(cols[-1]), int(rows[-1]), path.split("/")[-1])
    )

steady = []
for depth, boxes in sorted(per.items(), key=lambda kv: int(kv[0])):
    plain = [b for b in boxes if b[1] >= BANNER_ABOVE and b[3] <= BANNER_BELOW]
    arr = np.array([b[:4] for b in plain])
    steady.append(arr)
    print(
        "Depth%-2s frames=%-3d (%d without the banner)  x %d..%d  y %d..%d"
        % (depth, len(boxes), len(plain), arr[:, 0].min(), arr[:, 2].max(), arr[:, 1].min(), arr[:, 3].max())
    )
    for b in boxes:
        if b[1] < BANNER_ABOVE or b[3] > BANNER_BELOW:
            print("    device banner visible in", b[4], b[:4])

union = np.concatenate(steady)
print(
    "\nUnion over the frames without the banner: x %d..%d  y %d..%d  (%d x %d)"
    % (
        union[:, 0].min(),
        union[:, 2].max(),
        union[:, 1].min(),
        union[:, 3].max(),
        union[:, 2].max() - union[:, 0].min() + 1,
        union[:, 3].max() - union[:, 1].min() + 1,
    )
)
